# SQL-проект «Сервис для чтения книг»

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#SQL-запросы" data-toc-modified-id="SQL-запросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>SQL-запросы</a></span><ul class="toc-item"><li><span><a href="#Посчитаем-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитаем-сколько-книг-вышло-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитаем сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитаем количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой-книг" data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-книг-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Автор с самой высокой средней оценкой книг</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок." data-toc-modified-id="Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей, которые поставили больше 48 оценок.</a></span></li></ul></li><li><span><a href="#Итоговый-вывод" data-toc-modified-id="Итоговый-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Итоговый вывод</a></span></li></ul></div>

## Описание проекта

<div class="alert alert-info">
<font size="3", color = "black"><b> Цель проекта:  </b> <br/> проанализировать базу данных о книгах, издательствах, авторах и пользовательских обзорах, чтобы сформулировать ценностное предложение для нового продукта.

## Подготовка данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
import sqlalchemy as sa
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

#создаем функцию
def get_sql(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
  with engine.connect() as con:
    return pd.read_sql(sql=sa.text(query), con = con)


<div class="alert alert-info">
<font size="3", color = "black">
Рассмотрим таблицу с информацией о книгах

In [2]:
query = '''SELECT * FROM books LIMIT 10'''
get_sql(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

<div class="alert alert-info">
<font size="3", color = "black">
Рассмотрим таблицу с информацией об авторах

In [3]:
query = '''SELECT * FROM authors LIMIT 10'''
get_sql(query)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

<div class="alert alert-info">
<font size="3", color = "black">
Рассмотрим таблицу с информацией об издательствах

In [4]:
query = '''SELECT * FROM publishers LIMIT 10'''
get_sql(query)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

<div class="alert alert-info">
<font size="3", color = "black">
Рассмотрим таблицу с информацией о пользовательских оценках

In [5]:
query = '''SELECT * FROM ratings LIMIT 10'''
get_sql(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

<div class="alert alert-info">
<font size="3", color = "black">
Рассмотрим таблицу с информацией о пользовательских обзорах

In [6]:
query = '''SELECT * FROM reviews LIMIT 10'''
get_sql(query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

## SQL-запросы

### Посчитаем сколько книг вышло после 1 января 2000 года

In [7]:
query = '''SELECT 
    COUNT(DISTINCT book_id)
FROM 
    books
WHERE 
    publication_date > '2000-01-01';'''
print(get_sql(query))

   count
0    819


<div class="alert alert-info">
<font size="3", color = "black">
    После 1 января 2000 года вышло 819 книг.

### Для каждой книги посчитаем количество обзоров и среднюю оценку

In [8]:
query = '''WITH q1 AS (SELECT b.book_id, b.title, COUNT(rev.review_id) AS review_count
                                FROM books AS b
                                LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
                                GROUP BY b.book_id),
                         q2 AS (SELECT b.book_id,
                                      b.title,
                                      AVG(r.rating) AS avg_rating
                               FROM books AS b
                               LEFT JOIN ratings AS r ON b.book_id = r.book_id
                               GROUP BY b.book_id)
                     SELECT q1.book_id,
                            q1.title,
                            q1.review_count,
                            ROUND(q2.avg_rating, 2) AS avg_rating
                     FROM q1
                     INNER JOIN q2 ON q1.book_id = q2.book_id
                     GROUP BY q1.book_id, q1.title, q1.review_count, q2.avg_rating'''
print(get_sql(query))

     book_id                                              title  review_count  \
0        892  The Ultimate Hitchhiker's Guide to the Galaxy ...             3   
1        484                        Odd Thomas (Odd Thomas  #1)             4   
2        921  Think and Grow Rich: The Landmark Bestseller N...             3   
3        296                    Hard Eight (Stephanie Plum  #8)             3   
4         85                                             Anthem             4   
..       ...                                                ...           ...   
995      313              Heretics of Dune (Dune Chronicles #5)             2   
996      934          Traveling Mercies: Some Thoughts on Faith             2   
997      917                      The Yiddish Policemen's Union             2   
998      802                       The Moon is a Harsh Mistress             3   
999      479                             No Country for Old Men             4   

     avg_rating  
0        

<div class="alert alert-info">
<font size="3", color = "black">
Вывели таблицу, которая включает в себя айди книги, её название, количество пользовательских обзоров и среднюю оценку.

### Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [9]:
query = '''SELECT publishers.publisher, COUNT(*)
FROM publishers
JOIN books ON publishers.publisher_id = books.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher
ORDER BY COUNT(*) DESC
LIMIT 1;'''
print(get_sql(query))

       publisher  count
0  Penguin Books     42


<div class="alert alert-info">
<font size="3", color = "black">
    Наибольшее число книг толще 50 страниц выпустило издательство «Penguin Books»

### Автор с самой высокой средней оценкой книг

In [10]:
query = '''SELECT a.author, AVG(r.rating) AS avg_rating
FROM authors AS a
JOIN books AS b ON a.author_id = b.author_id
JOIN (
    SELECT book_id
    FROM ratings
    GROUP BY book_id
    HAVING COUNT(rating_id) >= 50
) AS q ON b.book_id = q.book_id
JOIN ratings AS r ON b.book_id = r.book_id
GROUP BY a.author
ORDER BY avg_rating DESC
LIMIT 1;'''
print(get_sql(query))

                       author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097


<div class="alert alert-info">
<font size="3", color = "black">
Автор с самой высокой средней оценкой книг — Дж. К. Роулинг.

###  Среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [11]:
query = '''SELECT AVG(reviews_count) AS avg_reviews_count
FROM (
    SELECT COUNT(review_id) AS reviews_count
    FROM reviews
    WHERE username IN (
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 48
    )
    GROUP BY username
) AS q;'''
print(get_sql(query))

   avg_reviews_count
0               24.0


<div class="alert alert-info">
<font size="3", color = "black">
Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: 24.

## Итоговый вывод

Был проведен анализ данных с помощью sql-запросов. Вот что удалось выяснить:
1. После 1 января 2000 года вышло 819 книг.
2. Для каждой книги посчитали количество обзоров и среднюю оценку
3. Наибольшее число книг толще 50 страниц выпустило издательство «Penguin Books»
4. Автор с самой высокой средней оценкой книг — Дж. К. Роулинг
5. Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: 24